### ISSUE TITLE
Unexpected warnings in Local Mode (chef-client -z) when cookbook_path has multiple entries

### ISSUE THREAD

**Issue by agh** [chef/chef#1778]
When you pass in multiple entries to `cookbook_path` using `chef-client -z` things get screwy.

## Key Info

 * Version: Chef Client 11.14.0.rc.2 (as shipped with ChefDK)
 * Platform: OSX 10.10 DP5

Seems to be reproducible on other platforms so I'm not sure this is specifically due to the above?

#### Attempted Flow

Very simply what I want is `chef-client -c etc/client.rb -z -o recipe[baa]` and to load `recipe[baa]` from `cookbooks/` and then have that depend on `recipe[sheep]` which is located in `site-cookbooks/`. Obviously you cannot Baa without having a Sheep.

#### Expectation

I would expect to be able to pass in multiple things to `cookbook_path` and to not result in that causing entries in `nodes/` (or other directories) to be evaluated more than once, and to not see warnings.

I've spelunnnked as much as possible but I'm getting lost in ChefFS and MultiplexedDir with regards to this bug, there's just too much metaprogramming going on there, it's significantly over my head. Everyone on IRC suggested @jkeiser is the most experienced at understanding this stuff.

## Observations / Steps to Reproduce

I have made the following observations which may help you nail down the specific bug.

#### Providing an invalid cookbook_path

If I update the entries in `client.rb` to specify */root/chef/cookbooks* and */root/chef/site-cookbooks* then the entire run fails, because chef-client tries to look in */root/chef/nodes* for the JSON data. 

```
stone:chef agh$ chef-client -c etc/client.rb -z -o recipe[baa]
Starting Chef Client, version 11.14.0.rc.2
[2014-08-07T22:34:31+01:00] ERROR: #<Errno::ENOENT: No such file or directory @ dir_s_mkdir - /root/chef/nodes>
/opt/chefdk/embedded/apps/chef/lib/chef/chef_fs/file_system/chef_repository_file_system_root_dir.rb:57:in `mkdir'
/opt/chefdk/embedded/apps/chef/lib/chef/chef_fs/file_system/chef_repository_file_system_root_dir.rb:57:in `block in create_child'
/opt/chefdk/embedded/apps/chef/lib/chef/chef_fs/file_system/chef_repository_file_system_root_dir.rb:55:in `each'
/opt/chefdk/embedded/apps/chef/lib/chef/chef_fs/file_system/chef_repository_file_system_root_dir.rb:55:in `create_child'
/opt/chefdk/embedded/apps/chef/lib/chef/chef_fs/chef_fs_data_store.rb:406:in `get_dir'
/opt/chefdk/embedded/apps/chef/lib/chef/chef_fs/chef_fs_data_store.rb:395:in `with_dir'
/opt/chefdk/embedded/apps/chef/lib/chef/chef_fs/chef_fs_data_store.rb:68:in `create'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/data_store/v1_to_v2_adapter.rb:56:in `block in create'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/data_store/v1_to_v2_adapter.rb:168:in `fix_exceptions'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/data_store/v1_to_v2_adapter.rb:55:in `create'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/rest_base.rb:92:in `create_data'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/endpoints/rest_list_endpoint.rb:30:in `post'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/rest_base.rb:29:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/rest_router.rb:23:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/server.rb:443:in `block in app'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/rack-1.5.2/lib/rack/handler/webrick.rb:60:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/rack-1.5.2/lib/rack/handler/webrick.rb:60:in `service'
/opt/chefdk/embedded/lib/ruby/2.1.0/webrick/httpserver.rb:138:in `service'
/opt/chefdk/embedded/lib/ruby/2.1.0/webrick/httpserver.rb:94:in `run'
/opt/chefdk/embedded/lib/ruby/2.1.0/webrick/server.rb:295:in `block in start_thread'

================================================================================
Chef encountered an error attempting to load the node data for "stone.local"
================================================================================

Unknown Server Error:
---------------------
The server had a fatal error attempting to load the node data.

Server Response:
----------------
Exception raised!  #<Errno::ENOENT: No such file or directory @ dir_s_mkdir - /root/chef/nodes>
/opt/chefdk/embedded/apps/chef/lib/chef/chef_fs/file_system/chef_repository_file_system_root_dir.rb:57:in `mkdir'
/opt/chefdk/embedded/apps/chef/lib/chef/chef_fs/file_system/chef_repository_file_system_root_dir.rb:57:in `block in create_child'
/opt/chefdk/embedded/apps/chef/lib/chef/chef_fs/file_system/chef_repository_file_system_root_dir.rb:55:in `each'
/opt/chefdk/embedded/apps/chef/lib/chef/chef_fs/file_system/chef_repository_file_system_root_dir.rb:55:in `create_child'
/opt/chefdk/embedded/apps/chef/lib/chef/chef_fs/chef_fs_data_store.rb:406:in `get_dir'
/opt/chefdk/embedded/apps/chef/lib/chef/chef_fs/chef_fs_data_store.rb:395:in `with_dir'
/opt/chefdk/embedded/apps/chef/lib/chef/chef_fs/chef_fs_data_store.rb:68:in `create'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/data_store/v1_to_v2_adapter.rb:56:in `block in create'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/data_store/v1_to_v2_adapter.rb:168:in `fix_exceptions'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/data_store/v1_to_v2_adapter.rb:55:in `create'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/rest_base.rb:92:in `create_data'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/endpoints/rest_list_endpoint.rb:30:in `post'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/rest_base.rb:29:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/rest_router.rb:23:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-zero-2.2/lib/chef_zero/server.rb:443:in `block in app'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/rack-1.5.2/lib/rack/handler/webrick.rb:60:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/rack-1.5.2/lib/rack/handler/webrick.rb:60:in `service'
/opt/chefdk/embedded/lib/ruby/2.1.0/webrick/httpserver.rb:138:in `service'
/opt/chefdk/embedded/lib/ruby/2.1.0/webrick/httpserver.rb:94:in `run'
/opt/chefdk/embedded/lib/ruby/2.1.0/webrick/server.rb:295:in `block in start_thread'

[2014-08-07T22:34:31+01:00] FATAL: Stacktrace dumped to /Users/agh/.chef/local-mode-cache/cache/chef-stacktrace.out
Chef Client failed. 0 resources updated in 2.006429 seconds
[2014-08-07T22:34:31+01:00] ERROR: 500 "Internal Server Error "
[2014-08-07T22:34:31+01:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

#### cookbook_path is valid and contains two entries

As described in the Attempted Flow, this has one cookbook in each directory and we pass in two directories to `cookbook_path` which is supposedly supported according to documentation.

```
# This setting enables SSL validation for HTTPS connections between
# chef-client and chef-server (but does not affect other HTTPS connections).
verify_api_cert true

cookbook_path [ "/Users/agh/src/agh/chef/cookbooks", "/Users/agh/src/agh/chef/site-cookbooks" ]
```

This does work but it emits an unexpected warning:

```
stone:chef agh$ chef-client -c etc/client.rb -z -o recipe[baa]
Starting Chef Client, version 11.14.0.rc.2
[2014-08-07T22:42:48+01:00] WARN: Child with name 'stone.local.json' found in multiple directories: /Users/agh/src/agh/chef/nodes/stone.local.json and /Users/agh/src/agh/chef/nodes/stone.local.json
[2014-08-07T22:42:48+01:00] WARN: Run List override has been provided.
[2014-08-07T22:42:48+01:00] WARN: Original Run List: []
[2014-08-07T22:42:48+01:00] WARN: Overridden Run List: [recipe[baa]]
resolving cookbooks for run list: ["baa"]
Synchronizing Cookbooks:
  - baa
  - sheep
Compiling Cookbooks...
Converging 0 resources
[2014-08-07T22:42:48+01:00] WARN: Skipping final node save because override_runlist was given

Running handlers:
Running handlers complete
Chef Client finished, 0/0 resources updated in 2.136351 seconds
```

The unexpected warning is `[2014-08-07T22:42:48+01:00] WARN: Child with name 'stone.local.json' found in multiple directories: /Users/agh/src/agh/chef/nodes/stone.local.json and /Users/agh/src/agh/chef/nodes/stone.local.json` and suggests that something is incorrectly making the assumption cookbook_path should be singular.

#### cookbook_path with three entries

I've now created another directory (`mkdir fake-cookbooks`) and added it to cookbook_path, however it is completely empty. This makes things get very shouty indeed.

```
# This setting enables SSL validation for HTTPS connections between
# chef-client and chef-server (but does not affect other HTTPS connections).
verify_api_cert true

cookbook_path [
  "/Users/agh/src/agh/chef/cookbooks",
  "/Users/agh/src/agh/chef/site-cookbooks",
  "/Users/agh/src/agh/chef/site-cookbooks" ]
```

Output from the convergence -

```
stone:chef agh$ chef-client -c etc/client.rb -z -o recipe[baa]
Starting Chef Client, version 11.14.0.rc.2
[2014-08-07T22:46:34+01:00] WARN: Child with name 'stone.local.json' found in multiple directories: /Users/agh/src/agh/chef/nodes/stone.local.json and /Users/agh/src/agh/chef/nodes/stone.local.json
[2014-08-07T22:46:34+01:00] WARN: Child with name 'stone.local.json' found in multiple directories: /Users/agh/src/agh/chef/nodes/stone.local.json and /Users/agh/src/agh/chef/nodes/stone.local.json
[2014-08-07T22:46:34+01:00] WARN: Run List override has been provided.
[2014-08-07T22:46:34+01:00] WARN: Original Run List: []
[2014-08-07T22:46:34+01:00] WARN: Overridden Run List: [recipe[baa]]
resolving cookbooks for run list: ["baa"]
[2014-08-07T22:46:34+01:00] WARN: Child with name 'sheep' found in multiple directories: /Users/agh/src/agh/chef/site-cookbooks/sheep and /Users/agh/src/agh/chef/site-cookbooks/sheep
[2014-08-07T22:46:34+01:00] WARN: Child with name 'sheep' found in multiple directories: /Users/agh/src/agh/chef/site-cookbooks/sheep and /Users/agh/src/agh/chef/site-cookbooks/sheep
[2014-08-07T22:46:34+01:00] WARN: Child with name 'sheep' found in multiple directories: /Users/agh/src/agh/chef/site-cookbooks/sheep and /Users/agh/src/agh/chef/site-cookbooks/sheep
[2014-08-07T22:46:34+01:00] WARN: Child with name 'sheep' found in multiple directories: /Users/agh/src/agh/chef/site-cookbooks/sheep and /Users/agh/src/agh/chef/site-cookbooks/sheep
[2014-08-07T22:46:34+01:00] WARN: Child with name 'sheep' found in multiple directories: /Users/agh/src/agh/chef/site-cookbooks/sheep and /Users/agh/src/agh/chef/site-cookbooks/sheep
[2014-08-07T22:46:34+01:00] WARN: Child with name 'sheep' found in multiple directories: /Users/agh/src/agh/chef/site-cookbooks/sheep and /Users/agh/src/agh/chef/site-cookbooks/sheep
[2014-08-07T22:46:34+01:00] WARN: Child with name 'sheep' found in multiple directories: /Users/agh/src/agh/chef/site-cookbooks/sheep and /Users/agh/src/agh/chef/site-cookbooks/sheep
[2014-08-07T22:46:34+01:00] WARN: Child with name 'sheep' found in multiple directories: /Users/agh/src/agh/chef/site-cookbooks/sheep and /Users/agh/src/agh/chef/site-cookbooks/sheep
[2014-08-07T22:46:34+01:00] WARN: Child with name 'sheep' found in multiple directories: /Users/agh/src/agh/chef/site-cookbooks/sheep and /Users/agh/src/agh/chef/site-cookbooks/sheep
[2014-08-07T22:46:34+01:00] WARN: Child with name 'sheep' found in multiple directories: /Users/agh/src/agh/chef/site-cookbooks/sheep and /Users/agh/src/agh/chef/site-cookbooks/sheep
[2014-08-07T22:46:34+01:00] WARN: Child with name 'sheep' found in multiple directories: /Users/agh/src/agh/chef/site-cookbooks/sheep and /Users/agh/src/agh/chef/site-cookbooks/sheep
[2014-08-07T22:46:34+01:00] WARN: Child with name 'sheep' found in multiple directories: /Users/agh/src/agh/chef/site-cookbooks/sheep and /Users/agh/src/agh/chef/site-cookbooks/sheep
Synchronizing Cookbooks:
  - baa
  - sheep
Compiling Cookbooks...
Converging 0 resources
[2014-08-07T22:46:34+01:00] WARN: Skipping final node save because override_runlist was given

Running handlers:
Running handlers complete
Chef Client finished, 0/0 resources updated in 2.091054 seconds
```

#### cookbook_path is valid and contains a single entry

If I move the `sheep` cookbook over, and adjust the configuration file -

```
# This setting enables SSL validation for HTTPS connections between
# chef-client and chef-server (but does not affect other HTTPS connections).
verify_api_cert true

cookbook_path [ "/Users/agh/src/agh/chef/cookbooks" ]
#cookbook_path [ "/Users/agh/src/agh/chef/cookbooks", "/Users/agh/src/agh/chef/site-cookbooks" ]
```

Then this causes everything to work as I expect, with no errors or warnings.

```
stone:chef agh$ chef-client -c etc/client.rb -z -o recipe[baa]
Starting Chef Client, version 11.14.0.rc.2
[2014-08-07T22:41:41+01:00] WARN: Run List override has been provided.
[2014-08-07T22:41:41+01:00] WARN: Original Run List: []
[2014-08-07T22:41:41+01:00] WARN: Overridden Run List: [recipe[baa]]
resolving cookbooks for run list: ["baa"]
Synchronizing Cookbooks:
  - baa
  - sheep
Compiling Cookbooks...
Converging 0 resources
[2014-08-07T22:41:41+01:00] WARN: Skipping final node save because override_runlist was given

Running handlers:
Running handlers complete
Chef Client finished, 0/0 resources updated in 1.908445 seconds
```

**Comment by kongslund** [comment]
Which PR is that? I am experiencing this issue with Chef 12.3.0 running in local mode.

**Comment by lantrix** [comment]
Same issue with chef-client 12.3.0 running in zero/local mode (`-z`)

**Comment by elijahchancey** [comment]
I see the same with 12.4.1 using zero/local mode and specifying chef_repo_path. I do not specify cookbook_path in my /etc/chef/client.rb file.

**Comment by kwhrtsk** [comment]
I found workaround.

In my environment, this warn message does not appear when .chef/knife.rb includes `node_path` configuration. FYI.

**Comment by stevendanna** [comment]
I think this will be fixed when f4ad4af6 is released

**Comment by lamont-granquist** [comment]
Hopefully fixed in 12.5.0

**Comment by thommay** [comment]
Closing since this should be fixed in master and the forthcoming 12.5 release. If that's not the case, please feel free to reopen.

**Comment by travishein** [comment]
I had come across this problem where when trying to use chef-zero, with Chef 12.4.3-1 at the time.
And read the comments above where it should be fixed in 12.5 release.
